examples/3d/render_to_texture.rs

```diff
@@ -0,0 +1,244 @@
+            usage: TextureUsages::TEXTURE_BINDING
+                | TextureUsages::COPY_DST
+                | TextureUsages::RENDER_ATTACHMENT,
+        },
+        ..Default::default()
+    };
+
+    // fill image.data with zeroes
+    image.resize(size);
+
+    let image_handle = images.set(RENDER_IMAGE_HANDLE, image);
+
+    let cube_handle = meshes.add(Mesh::from(shape::Cube { size: 4.0 }));
+    let cube_material_handle = materials.add(StandardMaterial {
+        base_color: Color::rgb(0.8, 0.7, 0.6),
+        reflectance: 0.02,
+        unlit: false,
+        ..Default::default()
+    });
+
+    // This specifies the layer used for the first pass, which will be attached to the first pass camera and cube.
+    let first_pass_layer = RenderLayers::layer(1);
+
+    // The cube that will be rendered to the texture.
+    commands
+        .spawn_bundle(PbrBundle {
+            mesh: cube_handle,
+            material: cube_material_handle,
+            transform: Transform::from_translation(Vec3::new(0.0, 0.0, 1.0)),
+            ..Default::default()
+        })
+        .insert(FirstPassCube)
+        .insert(first_pass_layer);
+
+    // Light
+    // NOTE: Currently lights are shared between passes - see https://github.com/bevyengine/bevy/issues/3462
+    commands.spawn_bundle(PointLightBundle {
+        transform: Transform::from_translation(Vec3::new(0.0, 0.0, 10.0)),
+        ..Default::default()
+    });
+
+    // First pass camera
+    let render_target = RenderTarget::Image(image_handle);
+    clear_color.insert(render_target.clone(), Color::WHITE);
+    let mut first_pass_camera = PerspectiveCameraBundle {
+        camera: Camera {
+            name: Some(FIRST_PASS_CAMERA.to_string()),
+            target: render_target,
+            ..Default::default()
+        },
+        transform: Transform::from_translation(Vec3::new(0.0, 0.0, 15.0))
+            .looking_at(Vec3::default(), Vec3::Y),
+        ..Default::default()
+    };
+    active_cameras.add(FIRST_PASS_CAMERA);
+
+    let camera_projection = &mut first_pass_camera.perspective_projection;
+    camera_projection.update(size.width as f32, size.height as f32);
+    first_pass_camera.camera.projection_matrix = camera_projection.get_projection_matrix();
+    first_pass_camera.camera.depth_calculation = camera_projection.depth_calculation();
+
+    commands
+        .spawn_bundle(first_pass_camera)
+        .insert(first_pass_layer);
+    // NOTE: omitting the RenderLayers component for this camera may cause a validation error:
+    //
+    // thread 'main' panicked at 'wgpu error: Validation Error
+    //
+    //    Caused by:
+    //        In a RenderPass
+    //          note: encoder = `<CommandBuffer-(0, 1, Metal)>`
+    //        In a pass parameter
+    //          note: command buffer = `<CommandBuffer-(0, 1, Metal)>`
+    //        Attempted to use texture (5, 1, Metal) mips 0..1 layers 0..1 as a combination of COLOR_TARGET within a usage scope.
```

Comment:
Odd. I'm not sure what this means. Do you have a backtrace?

Reply:
This happens when a texture is being rendered to and sampled from in the same frame. You need to either double buffer the texture and swap every frame or limit the usage using render layers like here.

Reply:
Here's a run with `BACKTRACE=full`:

```
$ RUST_BACKTRACE=full cargo run --example render_to_texture
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/examples/render_to_texture`
2021-12-29T00:42:32.457792Z  INFO bevy_render::renderer: AdapterInfo { name: "Apple M1 Max", vendor: 0, device: 0, device_type: DiscreteGpu, backend: Metal }
2021-12-29T00:42:32.710783Z ERROR wgpu::backend::direct: Handling wgpu errors as fatal by default
thread 'main' panicked at 'wgpu error: Validation Error

Caused by:
    In a RenderPass
      note: encoder = `<CommandBuffer-(0, 1, Metal)>`
    In a pass parameter
      note: command buffer = `<CommandBuffer-(0, 1, Metal)>`
    Attempted to use texture (8, 1, Metal) mips 0..1 layers 0..1 as a combination of COLOR_TARGET within a usage scope.

', /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2273:5
stack backtrace:
   0:        0x105de5248 - std::backtrace_rs::backtrace::libunwind::trace::hcf987238e0179083
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:        0x105de5248 - std::backtrace_rs::backtrace::trace_unsynchronized::h14d576ba2727cca2
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x105de5248 - std::sys_common::backtrace::_print_fmt::he2f4c6e9260671ee
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:67:5
   3:        0x105de5248 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h922ba3876a4c0b11
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:46:22
   4:        0x105dffae4 - core::fmt::write::h15b38023853d48c8
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/fmt/mod.rs:1149:17
   5:        0x105de1ec8 - std::io::Write::write_fmt::hc8d6ae95bf9a9dfe
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/io/mod.rs:1697:15
   6:        0x105de69f4 - std::sys_common::backtrace::_print::h6d92eb30e91656c5
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:49:5
   7:        0x105de69f4 - std::sys_common::backtrace::print::h979dc27da0f830e7
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:36:9
   8:        0x105de69f4 - std::panicking::default_hook::{{closure}}::h6ec26f90417d41e8
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:211:50
   9:        0x105de65d4 - std::panicking::default_hook::hb3d5f8274a134b7e
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:228:9
  10:        0x105de7034 - std::panicking::rust_panic_with_hook::h69d9e6b56bd395c4
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:606:17
  11:        0x105de6b80 - std::panicking::begin_panic_handler::{{closure}}::h4cf40d5b54276876
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:502:13
  12:        0x105de5714 - std::sys_common::backtrace::__rust_end_short_backtrace::h97ae71bd804cefeb
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:139:18
  13:        0x105de6ae8 - rust_begin_unwind
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:498:5
  14:        0x105e2816c - core::panicking::panic_fmt::h61e5a76b978edc22
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/panicking.rs:107:14
  15:        0x1051c8bc4 - wgpu::backend::direct::default_error_handler::h3a105ada710b3bd9
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2273:5
  16:        0x105203b94 - core::ops::function::Fn::call::h31b6080b895fbf30
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ops/function.rs:70:5
  17:        0x10529f58c - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h9693fc238e859aab
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/alloc/src/boxed.rs:1825:9
  18:        0x1051c893c - wgpu::backend::direct::ErrorSinkRaw::handle_error::hcbeb5ec9737b688b
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2259:17
  19:        0x1051c11ac - wgpu::backend::direct::Context::handle_error::ha1d04908a626db12
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:194:9
  20:        0x1051c820c - <wgpu::backend::direct::Context as wgpu::Context>::command_encoder_end_render_pass::h06212c76609179ac
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/backend/direct.rs:2010:13
  21:        0x1051eaeac - <wgpu::RenderPass as core::ops::drop::Drop>::drop::hc7f733f5207afacc
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.12.0/src/lib.rs:2890:13
  22:        0x104edfc94 - core::ptr::drop_in_place<wgpu::RenderPass>::h5c0d38da8ebe7129
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ptr/mod.rs:188:1
  23:        0x104ea195c - core::ptr::drop_in_place<bevy_render::render_phase::draw_state::TrackedRenderPass>::h74855d1efaac61b0
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ptr/mod.rs:188:1
  24:        0x104e93480 - <bevy_core_pipeline::main_pass_3d::MainPass3dNode as bevy_render::render_graph::node::Node>::run::h3db404f86423f778
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_core_pipeline/src/main_pass_3d.rs:89:9
  25:        0x104f87d78 - bevy_render::renderer::graph_runner::RenderGraphRunner::run_graph::hdbda2cb637778499
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_render/src/renderer/graph_runner.rs:178:17
  26:        0x104f87fec - bevy_render::renderer::graph_runner::RenderGraphRunner::run_graph::hdbda2cb637778499
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_render/src/renderer/graph_runner.rs:187:21
  27:        0x104f871d0 - bevy_render::renderer::graph_runner::RenderGraphRunner::run::h0420fe9c6901f0e1
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_render/src/renderer/graph_runner.rs:60:9
  28:        0x104ef4374 - bevy_render::renderer::render_system::he14bb2f38bc9d0e7
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_render/src/renderer/mod.rs:24:5
  29:        0x104ed50c4 - core::ops::function::FnMut::call_mut::h51ff4d09d8b58f1b
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ops/function.rs:150:5
  30:        0x104f796c0 - <bevy_ecs::system::exclusive_system::ExclusiveSystemFn<F> as bevy_ecs::system::exclusive_system::ExclusiveSystem>::run::hcbabd966d042d029
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_ecs/src/system/exclusive_system.rs:38:9
  31:        0x105c61dc4 - <bevy_ecs::schedule::stage::SystemStage as bevy_ecs::schedule::stage::Stage>::run::h149b6c5d59a7cc07
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_ecs/src/schedule/stage.rs:892:25
  32:        0x104fd55e0 - <bevy_render::RenderPlugin as bevy_app::plugin::Plugin>::build::{{closure}}::h1aa23f55c990a60e
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_render/src/lib.rs:262:17
  33:        0x105beb0c0 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hd3bf2b649537884b
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/alloc/src/boxed.rs:1825:9
  34:        0x105bf425c - bevy_app::app::App::update::hfb34b29be4b90704
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_app/src/app.rs:99:13
  35:        0x104846f60 - bevy_winit::winit_runner_with::{{closure}}::ha3b623af8933e4a7
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_winit/src/lib.rs:504:21
  36:        0x104868ec0 - <winit::platform_impl::platform::app_state::EventLoopHandler<T> as winit::platform_impl::platform::app_state::EventHandler>::handle_nonuser_event::{{closure}}::h1885a6f00c9260e9
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/app_state.rs:101:13
  37:        0x10486918c - winit::platform_impl::platform::app_state::EventLoopHandler<T>::with_callback::h2cf19d5d74cb7410
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/app_state.rs:79:13
  38:        0x104868da0 - <winit::platform_impl::platform::app_state::EventLoopHandler<T> as winit::platform_impl::platform::app_state::EventHandler>::handle_nonuser_event::hb6dd97a21ed9eeb0
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/app_state.rs:100:9
  39:        0x104879580 - winit::platform_impl::platform::app_state::Handler::handle_nonuser_event::h1807c342ddf27b60
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/app_state.rs:206:21
  40:        0x10487ae1c - winit::platform_impl::platform::app_state::AppState::cleared::ha2683d804edbb12c
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/app_state.rs:387:9
  41:        0x1048ad598 - winit::platform_impl::platform::observer::control_flow_end_handler::{{closure}}::h5799d630616ea1d3
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/observer.rs:180:21
  42:        0x1048ad3dc - winit::platform_impl::platform::observer::control_flow_handler::{{closure}}::h911eb68317557e2d
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/observer.rs:142:57
  43:        0x10488eef0 - std::panicking::try::do_call::h501ca9ea97765bc0
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:406:40
  44:        0x104893a70 - <unknown>
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/event_loop.rs:35:17
  45:        0x10488ee48 - std::panicking::try::h89e6b7e52ae60118
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:370:19
  46:        0x1048839ac - std::panic::catch_unwind::hcdb675aab2561fc0
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panic.rs:133:14
  47:        0x104892e24 - winit::platform_impl::platform::event_loop::stop_app_on_panic::h4d53d10a513284d2
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/event_loop.rs:229:11
  48:        0x1048ad200 - winit::platform_impl::platform::observer::control_flow_handler::h38d168aed27f2cd8
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/observer.rs:142:5
  49:        0x1048ad520 - winit::platform_impl::platform::observer::control_flow_end_handler::haf47b592e4254c14
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/observer.rs:175:9
  50:        0x18e883e4c - <unknown>
  51:        0x18e883c98 - <unknown>
  52:        0x18e8832f0 - <unknown>
  53:        0x18e882734 - <unknown>
  54:        0x197414f68 - <unknown>
  55:        0x197414cdc - <unknown>
  56:        0x197414a9c - <unknown>
  57:        0x1913d8ce0 - <unknown>
  58:        0x1913d7584 - <unknown>
  59:        0x1913c95a4 - <unknown>
  60:        0x1048d07b0 - <() as objc::message::MessageArguments>::invoke::h6adc70946e10abcc
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/objc-0.2.7/src/message/mod.rs:128:17
  61:        0x1048d1734 - objc::message::platform::send_unverified::hffd02c617bff251c
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/objc-0.2.7/src/message/apple/mod.rs:27:9
  62:        0x1048660a8 - objc::message::send_message::h2e3cb516ab95eeae
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/objc-0.2.7/src/message/mod.rs:178:5
  63:        0x1048660a8 - winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::{{closure}}::h13402eefd5b08b2c
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/event_loop.rs:188:22
  64:        0x104867b3c - objc::rc::autorelease::autoreleasepool::h1a7adf66d2186ec3
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/objc-0.2.7/src/rc/autorelease.rs:29:5
  65:        0x104865d60 - winit::platform_impl::platform::event_loop::EventLoop<T>::run_return::h4ffb73de1e3c0618
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/event_loop.rs:178:9
  66:        0x104866308 - winit::platform_impl::platform::event_loop::EventLoop<T>::run::h9fc45f926816ae1d
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/platform_impl/macos/event_loop.rs:157:9
  67:        0x104860fac - winit::event_loop::EventLoop<T>::run::hb03f1d684a9fe76e
                               at /Users/dusty/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.26.0/src/event_loop.rs:154:9
  68:        0x104846c34 - bevy_winit::run::hbbd9345c426ce8da
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_winit/src/lib.rs:171:5
  69:        0x10484a7d0 - bevy_winit::winit_runner_with::h99d26bfc0dbd3a4e
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_winit/src/lib.rs:513:9
  70:        0x10484a328 - bevy_winit::winit_runner::h84e8a4579b62d784
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_winit/src/lib.rs:211:5
  71:        0x104836a74 - core::ops::function::Fn::call::hf51b3574f6184163
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ops/function.rs:70:5
  72:        0x105beb078 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h0aabcb6b02375e4f
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/alloc/src/boxed.rs:1825:9
  73:        0x105bf4384 - bevy_app::app::App::run::h501fddd1b812cc22
                               at /Users/dusty/src/noumenal/crates/bevy/crates/bevy_app/src/app.rs:115:9
  74:        0x104824bfc - render_to_texture::main::h27cdae4b0127162c
                               at /Users/dusty/src/noumenal/crates/bevy/examples/3d/render_to_texture.rs:58:5
  75:        0x104821f84 - core::ops::function::FnOnce::call_once::h0ba1d9e0781990ee
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ops/function.rs:227:5
  76:        0x1048249e8 - std::sys_common::backtrace::__rust_begin_short_backtrace::h5774e667888abc55
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/sys_common/backtrace.rs:123:18
  77:        0x1048248d4 - std::rt::lang_start::{{closure}}::hfbddc14799dc1917
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/rt.rs:145:18
  78:        0x105de4be4 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h743b23775c597b1e
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/core/src/ops/function.rs:259:13
  79:        0x105de4be4 - std::panicking::try::do_call::h11b4ee64c16e3257
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:406:40
  80:        0x105de4be4 - std::panicking::try::habd571c9418c2fb2
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:370:19
  81:        0x105de4be4 - std::panic::catch_unwind::h14ae30993c89fa4d
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panic.rs:133:14
  82:        0x105de4be4 - std::rt::lang_start_internal::{{closure}}::h39341633e5562bf4
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/rt.rs:128:48
  83:        0x105de4be4 - std::panicking::try::do_call::hb38b25907952f248
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:406:40
  84:        0x105de4be4 - std::panicking::try::h63c96d16406a7a02
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panicking.rs:370:19
  85:        0x105de4be4 - std::panic::catch_unwind::ha30465df10c870f7
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/panic.rs:133:14
  86:        0x105de4be4 - std::rt::lang_start_internal::h6c393cd29ebc4567
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/rt.rs:128:20
  87:        0x1048248a0 - std::rt::lang_start::hb3e119d91ba52dab
                               at /rustc/efec545293b9263be9edfb283a7aa66350b3acbf/library/std/src/rt.rs:144:17
  88:        0x1048270cc - <unknown>
                               at /Users/dusty/src/noumenal/crates/bevy/examples/3d/render_to_texture.rs:242:9
```

Reply:
> This happens when a texture is being rendered to and sampled from in the same frame. You need to either double buffer the texture and swap every frame or limit the usage using render layers like here.

Ah, so this is a limitation of wgpu synchronisation? Vulkan and other modern APIs have synchronisation methods to allow waiting for work to be completed so that resources are in known states to then be able to reuse them or so, right? And wgpu has a model for synchronisation that is meant to 'just work' for all the common cases. I guess working around this limitation using double buffering or render layers is fine then. As double-buffering is more complex to implement, using the render layers probably makes sense for the example. I would be good to document the reason though as it's a bit of an unexpected gotcha in my opinion. :) Thanks for explaining @jakobhellermann !

Reply:
I've documented this in the example.